### ISSUE TITLE
Refactor database class to use PDO (+prepared statements) and to make it testable

### ISSUE THREAD

**Issue by andygrunwald** [lansuite/lansuite#183]
## Expected Behavior

All database queries are using native prepared statements by the [PDO MySQL extension/functionality of PHP](https://secure.php.net/manual/en/ref.pdo-mysql.php).

## Current Behavior

LanSuite uses a "kind of" prepared statements, but with a self written method.
See https://github.com/lansuite/lansuite/blob/master/inc/classes/class_db_mysql.php#L179 for details.

## Possible Solution

Inject a PDO MySQL object into the database class via the constructor (dependency injections).
With this, we can make this class testable, because we would be able to mock the object.

Furthermore, use native prepared statements like described in the PHP docs.
For this, we would need to rewrite every query to use the new syntax.

## Context

To use the PHP standard way would sense.
Other software engineers will have an easier time to get familiar with the code.
Prepared statements are the standard of the database interaction and well known to be secure.

### More resources

* [The 2018 Guide to Building Secure PHP Software: Database Interaction](https://paragonie.com/blog/2017/12/2018-guide-building-secure-php-software#secure-php-databases)
* [Are PDO prepared statements sufficient to prevent SQL injection? @ StackOverflow](https://stackoverflow.com/questions/134099/are-pdo-prepared-statements-sufficient-to-prevent-sql-injection/12202218#12202218)

## Your Environment

* Version used: current master (rev. fd582f8394fbda39bee2de8e8a98ffc8c5e2e99c)

**Comment by andygrunwald** [comment]
I found a related ticket in the old bugtracker. See http://lansuite.orgapage.de/index.php?mod=bugtracker&bugid=1171&ms_page=0

I had a small look and found zero Sleep-Queries in our Codebase.
I checked further and found a valid SQL Injection in the current LanSuite master and several older released version.
This explains the ticket in our old bugtracker.

Up to know i will not name how this injection can be triggered for security reasons (github don't have private issues). But enabling a proper database abstractions with prepared statement can avoid this.

**Comment by aj-gh** [comment]
Is going for _PDO MySQL_ a strict requirement, and if so, why?
_mysqli_ would already be there in the _DB_ class and I guess going for _mysqli_ prepared statements would be a rather minor initial change compared to a full-blown switchover to PDO. One could even slowly migrate any existing statements, possibly by providing any additionally needed functions of _mysqli_ via _DB_.

**Comment by andygrunwald** [comment]
> Is going for PDO MySQL a strict requirement, and if so, why?

No, it is not. I don't mind if we use _mysqli_ or PDO here, as long as we are using prepared statements.

**Comment by BenniGotchi** [comment]
I would like to push this topic and would offer to participate for the implementation - best case with some support from someone, that has more expirience in the PDO and a better overview about the Lansuite architecture than me.
From my perspective (perhaps I'm wrong with this) effort whise it doesn't make such a big difference, to go for PDO or mysqli - in principle each DB access has to be checked and modified in some way, so if PDO is the current standard approach and it would also provide some benefits regarding future testability, let's go for it ;)

Things that are not clear yet to me:

-  what to do with SQL statements, that do not get any parameters (from outside). E.g.
https://github.com/lansuite/lansuite/blob/fbf4fddc5076683f84baeb906c62c546501581cb/modules/party/plugins/home.php#L45
should a statement like this also be changed to predefined statements with a **prepare** and **execute** call?
I would tend to a 'yes', since it eliminates the risk, that a future change introduces silently a security issue?
- how to handle the %prefix% can this mechanism remain as it is today, or is there also the need to change?
> Up to know i will not name how this injection can be triggered for security reasons (github don't have private issues). But enabling a proper database abstractions with prepared statement can avoid this.

@andygrunwald : Would a 'internal' discussion here:
https://github.com/orgs/lansuite/teams/lansuite-developers
help, perhaps for this use case?

If you remember, we created this last year, but never used it so far...

**Comment by andygrunwald** [comment]
> should a statement like this also be changed to predefined statements with a prepare and execute call?

I would also for yes. Consistency makes it easier here. I agree with you.

> how to handle the %prefix% can this mechanism remain as it is today, or is there also the need to change?

We could potentially add this as a variable in the statement also. Like every other parameter. Needs to be tested.

> Would a 'internal' discussion here [...]

I don't see a need to discuss this internally. Or do we have secrets related to this topic? Let's discuss it here openly. Maybe someone come along and want to help. Who knows.

> If you remember, we created this last year, but never used it so far...

True. Right now, my time is quite limited to work on this.

**Comment by M4LuZ** [comment]
Gents,

I'm about to pick this up again as I've had a look at the related code and it appears that adapting the current custom code within `$db->qry` to use prepared statements appears doable with little work.
I would not worry about `%prefix%` here as it is not based on user input (besides the original setup stage), and the DB variables have to be set up for unit testing the class anyways.

Unless the vulnerability found is related to that?
+1 on using the LS-dev team for discussion of internal topics

**Comment by M4LuZ** [comment]
Spend a good amount of time to use already present separation of DML and input data to transform query strings to prepared statements.
There were a few issues I had to work around:
* `%plain%` - as that throws in additional SQL code and has to be treated separately
* variable type binding
* An issue with unbuffered resultsets causing error "Commands out of sync; you can't run this command now"
* ambigous return values of `stmt::get_result()` where false can both mean an empty result set or a command error

So far it appears that it runs fine for _some_ of the queries, while other parts of the code run into issues.
Queries logged look fine, so I suspect that there is still an issue with the return values that ends in a different reaction of the calling code.

I would appreciate if somebody takes a look from the side to see where I went wrong, code is in PR #488

**Comment by mrflobow** [comment]
Have you thought about implementing ORM instead of going prepared Statement route ? 
I know it might be a huge effort to switch to it , but maybe start with some .

You would be able to run unit test against them.

As I state in my other comment @M4LuZ the query fail because often mysqli_result is expected and if you use prepared statement it's another type.

So you need to be careful or create a fetch_array for prepared statement , probably in one single pub function. , maybe have two private function to split internally or write other wrappers like first(), count() etc..
